magefile.go

```diff
@@ -6,26 +6,188 @@ package main
+	}
+	shortCommitID = strings.TrimSpace(string(shortCommitIDBytes))
+
+	version = strings.TrimSpace(os.Getenv("VERSION"))
+	if len(version) == 0 {
+		version = shortCommitID
+	}
+
+	buildDate = fmt.Sprint(time.Now().UTC().Unix())
+	ldFlags = fmt.Sprintf(`-X %s/internal/version.Version=%s`+
+		`-X %s/internal/version.Branch=%s`+
+		`-X %s/internal/version.Commit=%s`+
+		`-X %s/internal/version.BuildDate=%s`,
+		module, version,
+		module, branch,
+		module, shortCommitID,
+		module, buildDate,
+	)
+	return nil
+}
+
+// Builds binaries from /cmd directory.
+func (Build) cmd(cmd, goos, goarch string) error {
+	mg.Deps(Build.init)
+
+	env := map[string]string{
+		"GOFLAGS":     "",
+		"CGO_ENABLED": "0",
+		"LDFLAGS":     ldFlags,
+	}
+
+	bin := path.Join("bin", cmd)
+	if len(goos) != 0 && len(goarch) != 0 {
+		// change bin path to point to a sudirectory when cross compiling
+		bin = path.Join("bin", goos+"_"+goarch, cmd)
+		env["GOARGS"] = fmt.Sprintf("GOOS=%s GOARCH=%s", goos, goarch)
+	}
+
+	if err := sh.RunWithV(
+		env,
+		"go", "build", "-v", "-o", bin, "./cmd/"+cmd+"/main.go",
+	); err != nil {
+		return fmt.Errorf("compiling cmd/%s: %w", cmd, err)
+	}
+	return nil
+}
+
+// Default build target for CI/CD
+func (Build) All() {
+	mg.Deps(
+		mg.F(Build.cmd, "addon-operator-manager", "linux", "amd64"),
+		mg.F(Build.cmd, "addon-operator-webhook", "linux", "amd64"),
+		mg.F(Build.cmd, "api-mock", "linux", "amd64"),
+		mg.F(Build.cmd, "mage", "", ""),
+	)
+}
+
+// Builds the docgen internal tool
+func (Build) Docgen() {
+	mg.Deps(mg.F(Build.cmd, "docgen", "", ""))
+}
+
+// Code Generators
+// ---------------
+type Generate mg.Namespace
+
+func (Generate) All() {
+	mg.Deps(
+		Generate.code,
+		Generate.docs,
+	)
+}
+
+func (Generate) code() error {
+	mg.Deps(Dependency.ControllerGen)
+
+	manifestsCmd := exec.Command("controller-gen",
+		"crd:crdVersions=v1", "rbac:roleName=addon-operator-manager",
+		"paths=./...", "output:crd:artifacts:config=../config/deploy")
+	manifestsCmd.Dir = workDir + "/apis"
+	if err := manifestsCmd.Run(); err != nil {
+		return fmt.Errorf("generating kubernetes manifests: %w", err)
+	}
+
+	// code gen
+	codeCmd := exec.Command("controller-gen", "object", "paths=./...")
+	codeCmd.Dir = workDir + "/apis"
+	if err := codeCmd.Run(); err != nil {
+		return fmt.Errorf("generating deep copy methods: %w", err)
+	}
+
+	// patching generated code to stay go 1.16 output compliant
+	// https://golang.org/doc/go1.17#gofmt
+	// @TODO: remove this when we move to go 1.17"
+	// otherwise our ci will fail because of changed files"
+	// this removes the line '//go:build !ignore_autogenerated'"
+	if err := sh.Run("find", ".", "-name", "zz_generated.deepcopy.go", "-exec", "sed", "-i", `/\/\/go:build !ignore_autogenerated/d`, "{}", ";"); err != nil {
+		return fmt.Errorf("removing go:build annotation: %w", err)
+	}
+
+	return nil
+}
+
+func (Generate) docs() error {
+	mg.Deps(Build.Docgen)
+
+	return sh.Run("./hack/docgen.sh")
+}
```

Comment:
Just out of curiosity: Any reason to not expose this target unlike previously exposed `make generate-docs` target?

Reply:
nvm, I see the point in not needing to expose the `generate-code` and `generate-docs` targets considering them not being used separately anywhere 😄